[mlir][vector] Restrict DropInnerMostUnitDimsTransfer{Read|Write}

banach-space opened the pull request:
Restrict `DropInnerMostUnitDimsTransfer{Read|Write}` so that it fails when
one of the indices to be dropped could be != 0 and "out of bounds":

```mlir
func.func @negative_example(%arg0: memref<16x1xf32>, %arg1: vector<8x1xf32>, %idx_1: index, %idx_2: index) {
  vector.transfer_write %arg1, %arg0[%idx_1, %idx_2] {in_bounds = [true, false]} : vector<8x1xf32>, memref<16x1xf32>
  return
}
```

This is an edge case that could represent an out-of-bounds access,
though that will depend on the actual value of %i. Importantly, without
this change it would be transformed as follows:

```mlir
func.func @negative_example(%arg0: memref<16x1xf32>, %arg1: vector<8x1xf32>, %arg2: index, %arg3: index) {
  %subview = memref.subview %arg0[0, 0] [16, 1] [1, 1] : memref<16x1xf32> to memref<16xf32, strided<[1]>>
  %0 = vector.shape_cast %arg1 : vector<8x1xf32> to vector<8xf32>
  vector.transfer_write %0, %subview[%arg2] {in_bounds = [true]} : vector<8xf32>, memref<16xf32, strided<[1]>>
  return
}
```

This is incorrect - `%idx_2` is ignored and the "out of bounds" flags is
not propagated. Hence the extra restriction to avoid such cases.

NOTE: This is a follow-up for: #94904

banach-space committed:
[mlir][vector] Restrict DropInnerMostUnitDimsTransferWrite

Restrict `DropInnerMostUnitDimsTransferWrite` so that it fails when one
of the indices to be dropped could be != 0, e.g.

```mlir
func.func @negative_example(
    %arg0: memref<16x1xf32>,
    %arg1: vector<8x1xf32>,
    %idx_1: index,
    %idx_2: index) {

  %c0 = arith.constant 0 : index
  vector.transfer_write %arg1, %arg0[%idx_1, %idx_2] {in_bounds = [true, true]} : vector<8x1xf32>, memref<16x1xf32>
  return
}
```

This is an edge case that could represent an out-of-bounds access,
though that will depend on the actual value of `%i`. Importantly,
_without this change_ it would be transformed as follows:
```mlir
func.func @negative_example(
    %arg0: memref<16x1xf32>,
    %arg1: vector<8x1xf32>,
    %idx_1: index,
    %idx_2: index) {

  %subview = memref.subview %arg0[0, 0] [16, 1] [1, 1] : memref<16x1xf32> to memref<16xf32, strided<[1]>>
  %0 = vector.shape_cast %arg1 : vector<8x1xf32> to vector<8xf32>
  vector.transfer_write %0, %subview[%idx_1] {in_bounds = [true]} : vector<8xf32>, memref<16xf32, strided<[1]>>
  return
}
```

This is incorrect - `%idx_2` is ignored. Hence the extra restriction to
avoid such cases.

NOTE: This PR is limited to `vector.transfer_write`. Similar patch for
`vector.transfer_read`: #94904

nujaa commented on mlir/test/Dialect/Vector/vector-transfer-collapse-inner-most-dims.mlir:
```diff
@@ -367,6 +367,33 @@ func.func @contiguous_inner_most_dynamic_outer_scalable_inner_dim(%a: index, %b:
 
 // -----
 
+func.func @contiguous_inner_most_non_zero_idxs(%arg0: memref<16x1xf32>, %arg1: vector<8x1xf32>, %i: index) {
```
What additional feature does this test show compared to `contiguous_inner_most_dynamic_outer` ?

banach-space committed:
fixup! [mlir][vector] Restrict DropInnerMostUnitDimsTransferWrite

Remove duplicate test

MacDue commented:
I'm not convinced that the transform you show in the description is invalid. 

```
func.func @negative_example(
    %arg0: memref<16x1xf32>,
    %arg1: vector<8x1xf32>,
    %idx_1: index,
    %idx_2: index) {
  vector.transfer_write %arg1, %arg0[%idx_1, %idx_2] {in_bounds = [true, true]} : vector<8x1xf32>, memref<16x1xf32>
  return
}
```
Looking at this there's no mask and both dims are marked as in_bounds. Therefore `%idx_2` _must_ be 0, as the in_bounds attribute "specifies for every vector dimension if the transfer is **guaranteed** to be within the source bounds". 

So the program is invalid (and I think it invokes UB) if `%idx_2` is not zero (as it breaks that guarantee).

If `%idx_2` was marked as out-of-bounds (mask or no mask), I believe knowing that the dimension is 1 then it means the write can probably be optimized-out. To avoid the out-bounds-write would mean masking out the entire unit dimension, which means the write would be a no-op.

banach-space committed:
fixup! fixup! [mlir][vector] Restrict DropInnerMostUnitDimsTransferWrite

Allow non-zero indices when in-bounds

banach-space commented:
> Looking at this there's no mask and both dims are marked as in_bounds. Therefore `%idx_2` _must_ be 0, as the in_bounds attribute "specifies for every vector dimension if the transfer is **guaranteed** to be within the source bounds".

That's a very good point, thanks for taking the time to analyse this.

> So the program is invalid (and I think it invokes UB) if `%idx_2` is not zero (as it breaks that guarantee).
> 
> If `%idx_2` was marked as out-of-bounds (mask or no mask), I believe knowing that the dimension is 1 then it means the write can probably be optimized-out. To avoid the out-bounds-write would mean masking out the entire unit dimension, which means the write would be a no-op.

👍🏻 For that I'd like to clarify/formalise the interplay between masks and `in_bounds` first. 

> So, I think you just need to check the `in-bounds` attribute is true for the dimensions that you're going to drop.

Agreed, thanks! That's already updated (see the latest commit). I have also added a few tests to make the distinct cases very clear. Also, while `in_bounds` is optional, the code is a bit verbose. But we are making `in_bounds` mandatory:
* https://github.com/llvm/llvm-project/pull/97049

MacDue commented on mlir/lib/Dialect/Vector/Transforms/VectorTransforms.cpp:
```diff
@@ -1394,6 +1394,33 @@ class DropInnerMostUnitDimsTransferWrite
+    // index is "0". This could be added as a folder.
+    // TODO: Deal with 3. by e.g. proppaging the "out of bounds" flag to other
+    // dims.
+    bool indexOutOfBounds = true;
+    if (writeOp.getInBounds())
+      indexOutOfBounds = llvm::any_of(
+          llvm::zip(writeOp.getInBounds()->getValue().take_back(dimsToDrop),
+                    writeOp.getIndices().take_back(dimsToDrop)),
+          [](auto zipped) {
+            auto inBounds = cast<BoolAttr>(std::get<0>(zipped)).getValue();
+            auto nonZeroIdx = !isZeroIndex(std::get<1>(zipped));
+            return !inBounds && nonZeroIdx;
+          });
+    else
+      indexOutOfBounds = !llvm::all_of(
+          writeOp.getIndices().take_back(dimsToDrop), isZeroIndex);
+    if (indexOutOfBounds)
+      return failure();
```
This looks much more complex than I'd expect. I don't see the need to check the indices at all. 

Also, dropping the dim if the index is zero and marked as out-of-bounds does not seem valid. If index zero is out-of-bounds, then we can't safely write to that unit dimension (and the write probably should be a no-op). 

Why not just:
```c++
auto inBounds = writeOp.getInBoundsValues();
auto droppedInBounds = ArrayRef<bool>(inBounds).take_back(dimsToDrop);
if (llvm::is_contained(droppedInBounds, false)) 
  return failure();
```

MacDue replied:
(or maybe just):
```c++
if (writeOp.hasOutOfBoundsDim())
  return failure();
```
out-of-bounds dims are a edge-case (that I've not really seen used in practice, so are probably not worth worrying about).

MacDue replied:
Side note: If index 0 and in-bounds = false for a unit-dim actually means in-bounds, that should be an xferOp fold (not something everyone that looks at in-bounds values needs to interpret).

banach-space replied:
>  dropping the dim if the index is zero and marked as out-of-bounds does not seem valid. If index zero is out-of-bounds, then we can't safely write to that unit dimension

As you observed further down:

> If index 0 and in-bounds = false for a unit-dim actually means in-bounds = true

:) So:
1.  if the index is == 0 then `in_bounds` is effectively irrelevant (safe to collapse)
2. if index is != 0, but `in_bounds = true`, the index is effectively ==0 and  (safe to collapse)
3. if index != 0 and `in_bounds = false`, bail out.

Note that once `in_bounds` is mandatory, I will be able to simply the above as:
```cpp
      if(llvm::any_of(
          llvm::zip(writeOp.getInBounds()->getValue().take_back(dimsToDrop),
                    writeOp.getIndices().take_back(dimsToDrop)),
          [](auto zipped) {
            auto inBounds = cast<BoolAttr>(std::get<0>(zipped)).getValue();
            auto nonZeroIdx = !isZeroIndex(std::get<1>(zipped));
            return !inBounds && nonZeroIdx;
          }))
      return failure();
```

> out-of-bounds dims are an edge-case

Not sure we can say that - the default for `in_bounds` is "out of bounds".

banach-space replied:
> It's an actual fold too (foldTransferInBoundsAttribute, so it's running pretty much all the time).

I knew about the folder, but incorrectly assumed that the "folder" wasn't guaranteed to be run before the pattern. I was wrong: https://github.com/llvm/llvm-project/blob/1ed84a862f9ce3c60251968f23a5405f06458975/mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h#L107-L108

> /// Also performs folding and simple dead-code elimination before attempting to
/// match any of the provided patterns.

So:
> Why not just:
> 
> auto inBounds = writeOp.getInBoundsValues();
> auto droppedInBounds = ArrayRef<bool>(inBounds).take_back(dimsToDrop);
> if (llvm::is_contained(droppedInBounds, false)) 
>  return failure();

Yeah, that's perfectly sufficient and covers all the cases. Thanks!

nujaa replied:
> Not sure we can say that - the default for in_bounds is "out of bounds".
I think this is definitely safer.

banach-space committed:
fixup! fixup! fixup! [mlir][vector] Restrict DropInnerMostUnitDimsTransferWrite

Simplify as per Ben's suggestion. Given that the Op folder is guaranteed to run before the pattern, we can safely assume that the in_bounds attribute already contains all the info that we need.

MacDue reviewed (approved):
LGTM (though I think #94904 should be updated like this too :))

MacDue commented:
(also PR description needs updating)

banach-space committed:
fixup! fixup! fixup! [mlir][vector] Restrict DropInnerMostUnitDimsTransferWrite

Extend to xfer_read

banach-space commented:
>  I think https://github.com/llvm/llvm-project/pull/94904 should be updated like this too :)

Done

> (also PR description needs updating)

Done 

@nujaa , any further suggestions?

nujaa commented:
LGTM 👍
